tidb-binlog/driver/reader/kafka_consumer/kafka_go.go

```diff
@@ -0,0 +1,227 @@
+				Offset: kmsg.Offset,
+			}
+			consumerChan <- msg
+		}
+	}
+}
+
+// SeekOffsetFromTS implements Consumer.SeekOffsetFromTS
+func (k *KafkaGO) SeekOffsetFromTS(ts int64, topic string, partitions []int32) ([]int64, error) {
+	if len(partitions) == 0 {
+		pts, err := k.conn.ReadPartitions(topic)
+		if err != nil {
+			log.Error("get partitions from topic failed", zap.String("topic", topic), zap.Error(err))
+			return nil, errors.Trace(err)
+		}
+		for _, pt := range pts {
+			partitions = append(partitions, int32(pt.ID))
+		}
+	}
+	return k.seekOffsets(topic, partitions, ts)
+}
+
+// seekOffsets returns all valid offsets in partitions
+func (k *KafkaGO) seekOffsets(topic string, partitions []int32, pos int64) ([]int64, error) {
+	offsets := make([]int64, len(partitions))
+	for _, partition := range partitions {
+		start, err := k.conn.ReadFirstOffset()
+		if err != nil {
+			return nil, errors.Trace(err)
+		}
+
+		end, err := k.conn.ReadLastOffset()
+		if err != nil {
+			return nil, errors.Trace(err)
+		}
+
+		log.Info("seek offsets in",
+			zap.String("topic", topic),
+			zap.Int32("partition", partition),
+			zap.Int64("start", start),
+			zap.Int64("end", end),
+			zap.Int64("target ts", pos))
+
+		offset, err := k.seekOffset(topic, partition, start, end-1, pos)
+		if err != nil {
+			return nil, errors.Trace(err)
+		}
+
+		log.Info("seek offset success", zap.Int64("offset", offset), zap.Int64("target ts", pos))
+		offsets[partition] = offset
+	}
+
+	return offsets, nil
+}
+
+func (k *KafkaGO) seekOffset(topic string, partition int32, start int64, end int64, ts int64) (offset int64, err error) {
+	startTS, err := k.getTSAtOffset(topic, partition, start)
+	if err != nil {
+		err = errors.Trace(err)
+		return
+	}
+
+	if ts < startTS {
+		log.Warn("given ts is smaller than oldest message's ts, some binlogs may lose", zap.Int64("given ts", ts), zap.Int64("oldest ts", startTS))
+		offset = start
+		return
+	} else if ts == startTS {
+		offset = start + 1
+		return
+	}
+
+	for start < end {
+		mid := (end-start)/2 + start
+		var midTS int64
+		midTS, err = k.getTSAtOffset(topic, partition, mid)
+		if err != nil {
+			err = errors.Trace(err)
+			return
+		}
+
+		if midTS <= ts {
+			start = mid + 1
+		} else {
+			end = mid
+		}
+	}
+
+	var endTS int64
+	endTS, err = k.getTSAtOffset(topic, partition, end)
+	if err != nil {
+		err = errors.Trace(err)
+		return
+	}
+
+	if endTS <= ts {
+		return end + 1, nil
+	}
+
+	return end, nil
+}
+
+func (k *KafkaGO) getTSAtOffset(topic string, partition int32, offset int64) (ts int64, err error) {
+	log.Debug("start consumer on kafka",
+		zap.String("topic", topic),
+		zap.Int32("partition", partition),
+		zap.Int64("offset", offset))
+
+	_, err = k.conn.Seek(offset, 1)
+	if err != nil {
+		err = errors.Trace(err)
+		return
+	}
+
+	msg, err := k.conn.ReadMessage(10e6)
+	ts, err = getTSFromMSG(k.ConsumerType(), &KafkaMsg{
+		Offset: msg.Offset,
+		Value:  msg.Value,
+	})
+
+	if err == nil {
```

Comment:
Should also log if there's any error?

Reply:
if we got error here, this error trace already return to reader